### ISSUE TITLE
Investigate Gradle composite builds

### ISSUE THREAD

**Issue by cwardgar** [Unidata/thredds#667]
https://blog.gradle.org/introducing-composite-builds

> Splitting Monoliths
>
> Organizations that want to avoid the integration pains of multiple repositories tend to use a “monorepo”—a repository containing all projects, often including their dependencies and necessary tools. The upside is that all code is in one place and downstream breakages become visible immediately. But this convenience can come at the cost of productivity: a given developer will usually work only on a small part of a monorepo, but will still be forced to build all upstream projects, and that can mean a lot of waiting and wasted time. Likewise, importing large monorepo projects into an IDE often results in an unresponsive and overwhelming experience.
> 
> With composite builds, you can break your monorepo up into several independent builds within the same repository. Developers can work with the individual builds to get fast turnarounds or work with the whole composite when they want to ensure that everything still plays well together.

An important step of reducing the technical debt of the thredds project is breaking up its monolithic structure. Composite builds would allow us to split some of the subprojects off into their own discrete projects, and then be able to work on them without the typical pain you'd have if you did that. Good candidates are `bufr`, `grib`, `dap4`, `ui`, etc.

**Comment by cwardgar** [comment]
A major complication here are tests, which often straddle project boundaries.

**Comment by DennisHeimbigner** [comment]
In order to do this, we need a good dependency diagram.
It is for dependency issues that e.g. opendap and dap4 are spread
across multiple modules (e.g. opendap and tds).
Ideally (IMO) we should have every IOSP, every subclass of netcdfdataset
and every output protocol in their own project; this is currently
difficult because of dependency issues. Fixing this is not simple
with the current thredds architecture and modules.

**Comment by cwardgar** [comment]
Agreed. We'd need to think long and hard about good interfaces between modules and probably move to OSGi or Jigsaw. Definitely worth doing though.
